### PR TITLE
Add connection abstraction to devp2p networking

### PR DESCRIPTION
**Pull request by pipermerriam**
Extracted from #684 

Builds on #887

### What was wrong?

As a continuation of #684 we previously implemented the `Multiplexer` and new facilities to handle the P2P handshake under `p2p.handshake`.  Now we need something *active* (aka a Service) to manage the multiplexed connection and allow multiple consumers for each protocol or command.

### How was it fixed?

The implements a `Connection` service (which is implemented but not yet used by the `p2p.BasePeer`.  This service houses a `Multiplexer` and exposes the following APIs.

1. Mirrors access to the base `p2p` protocol from the `Multiplexer`.
2. Mirrors access to the `remote` from the `Multiplexer`
3. Meta-data about the remote connection like their public key, client version string, listen port, etc.
4. Meta-data about the handshake via the handshake receipts.
5. APIs for registering handlers for individual Commands or full Protocols.  This can be thought of as similar to the existing peer subscription API.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Clean up commit history

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)
- [x] Add entry to the [release notes](https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

#### Cute Animal Picture

extra-large-dog-breeds-great-dane

### CODE REVIEW

**Comment by pipermerriam** [comment]
Blocked by #887

File: p2p/connection.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    #
+    # Subscriptions/Handler API
+    #
+    async def _feed_protocol_handlers(self, protocol: ProtocolAPI) -> None:
+        # do not start consuming from the protocol stream until
+        # `start_protocol_streams` has been called and the multiplexer is
+        # active.
+        try:
+            await asyncio.wait_for(asyncio.gather(self._handlers_ready.wait()), timeout=10)
+        except TimeoutError:
```
Comment (pipermerriam) [inline p2p/connection.py]: I think this actually has to be `asyncio.TimeoutError` because it isn't the same as the built-ins `TimeoutError`

File: p2p/connection.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+            self.logger.info('Timedout waiting for handler ready signal')
+            raise TimeoutError(
+                "The handlers ready event was never set.  Ensure that "
+                "`Connection.start_protocol_streams()` is being called"
+            )
+
+        # we don't need to use wait_iter here because the multiplexer does it
+        # for us.
+        async for cmd, msg in self._multiplexer.stream_protocol_messages(protocol):
+            self.logger.debug('Handling command: %s', type(cmd))
```
Comment (pipermerriam) [inline p2p/connection.py]: probably too noisy.

File: p2p/tools/factories/multiplexer.py [hunk 3]
```diff
@@ -46,15 +60,22 @@ def MultiplexerPairFactory(*,
         in zip(protocol_types, cmd_id_offsets)
     )
 
-    alice_p2p_protocol = P2PProtocol(alice_transport, 0, snappy_support)
+    p2p_protocol_class: Type[BaseP2PProtocol]
+
+    if snappy_support:
+        p2p_protocol_class = P2PProtocol
+    else:
+        p2p_protocol_class = P2PProtocolV4
```
Comment (pipermerriam) [inline p2p/tools/factories/multiplexer.py]: These changes can get extracted into a separate PR.

**Comment by pipermerriam** [comment]
Requested review from both @lithp and @carver just to get both of you to look at it.  I think Jason has seen this code before when I was presenting it in a preliminary fashion, but Brian has reviewed some of the recent pull requests in this feature line.

File: p2p/abc.py [hunk 4]
```diff
@@ -321,3 +324,77 @@ def stream_protocol_messages(self,
     #
     def multiplex(self) -> AsyncContextManager[None]:
         ...
+
+
+class HandlerSubscriptionAPI:
```
Comment (lithp) [inline p2p/abc.py]: Some alternatives:
- Making this more abstract: Calling it `Cancellable` might make it more clear what you can do with these objects.
- Making it more concrete: mypy kind of gets in the way but this function also could return the function directly, first class functions are pretty dope! `CancelFn = Callable[[], None]`
- It doesn't look like these subscriptions are ever used, I bet we could get away with not providing a way to cancel subscriptions for quite a while.

File: p2p/connection.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+class HandlerSubscription(HandlerSubscriptionAPI):
+    def __init__(self, remove_fn: Callable[[], Any]) -> None:
+        self._remove_fn = remove_fn
+
+    def cancel(self) -> None:
+        self._remove_fn()
+
+
+class Connection(ConnectionAPI, BaseService):
+    disconnect_reason: DisconnectReason = None
```
Comment (lithp) [inline p2p/connection.py]: I don't think this attribute is ever set or read from

File: p2p/connection.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+)
+from p2p.service import BaseService
+from p2p.p2p_proto import (
+    BaseP2PProtocol,
+)
+from p2p.typing import Capabilities, Payload
+
+
+TProtocol = TypeVar('TProtocol', bound=ProtocolAPI)
+TReceipt = TypeVar('TReceipt', bound=HandshakeReceipt)
```
Comment (lithp) [inline p2p/connection.py]: Neither of these are used anywhere

File: p2p/connection.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+        self._command_handlers = collections.defaultdict(set)
+
+        # An event that controls when the connection will start reading from
+        # the individual multiplexed protocol streams and feeding handlers.
+        # This ensures that the connection does not start consuming messages
+        # before all necessary handlers have been added
+        self._handlers_ready = asyncio.Event()
+
+    def start_protocol_streams(self) -> None:
+        self._handlers_ready.set()
```
Comment (lithp) [inline p2p/connection.py]: Why not hold off on calling `_run` until all the correct handlers have been added?

File: p2p/connection.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        return not self.is_dial_out
+
+    @property
+    def remote(self) -> NodeAPI:
+        return self._multiplexer.remote
+
+    async def _run(self) -> None:
+        async with self._multiplexer.multiplex():
+            for protocol in self._multiplexer.get_protocols():
+                self.run_daemon_task(self._feed_protocol_handlers(protocol))
```
Comment (lithp) [inline p2p/connection.py]: The multiplexer goes through a lot of effort to disaggregate these streams, and then it spends most of its life here in `Connection` where those streams are brought back together again! It might simplify things for the handshaking code could use something less general than the full `Multiplexer`, and for `Connection` to take a transport directly.

File: p2p/connection.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+    @property
+    def remote_public_key(self) -> keys.PublicKey:
+        return keys.PublicKey(self._devp2p_receipt.remote_public_key)
+
+    @property
+    def client_version_string(self) -> str:
+        return self._devp2p_receipt.client_version_string
+
+    @property
+    def safe_client_version_string(self) -> str:
```
Comment (lithp) [inline p2p/connection.py]: I'm not sure where this will be used so maybe this is bad feedback, but you probably still want to make the string printable, even if it's under 256 characters!

There's also no test for this behavior! If you drop the lines which checks for isprintable and always return the string (never calling `repr`) the tests still pass.

File: p2p/tools/factories/connection.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+from p2p.service import run_service
+
+from .cancel_token import CancelTokenFactory
+from .multiplexer import MultiplexerPairFactory
+from .p2p_proto import DevP2PHandshakeParamsFactory
+from .transport import MemoryTransportPairFactory
+
+
+@asynccontextmanager
+async def ConnectionPairFactory(*,
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply p2p/tools/factories/connection.py]: Yes but that happens in the next pull request #962 .

File: p2p/abc.py (hunk 4 above)
Comment (pipermerriam) [reply p2p/abc.py]: In a subsequent PR I'm expanding this to be a context manager so that it cancels the handler on exit so I'm going to keep it as-is since that context management API proves useful for things that have a shorter lifecycle than the connection and thus need to have their handlers reliably de-registered.

File: p2p/connection.py (hunk 5 above)
Comment (pipermerriam) [reply p2p/connection.py]: will remove.  This is a carryover from when I had a `Connection.disconnect` method but I removed it late in the process because I wanted to keep the `Connection` *pure* and devoid of any actual business logic.

File: p2p/connection.py (hunk 7 above)
Comment (pipermerriam) [reply p2p/connection.py]: In theory that would work, except that it ends up working out that the places where the handlers get registered isn't at the right level to also be in charge of running the connection (the `Connection` running part happens at a level higher than the `Peer` and so the peer can't be responsible for running it).

This mechanism is pretty simple and it's working well but I'm happy to re-visit it if something better shows up later.

File: p2p/connection.py (hunk 8 above)
Comment (pipermerriam) [reply p2p/connection.py]: I think you might be right that we could adjust the internal `Connection` to just directly use `p2p.multiplexer.stream_transport_messages` and only use the multiplexer during the handshake.  I'm gonna hold off on making this change for now but good suggestion (I'm gonna open an issue to track)

File: p2p/connection.py (hunk 9 above)
Comment (pipermerriam) [reply p2p/connection.py]: :+1: will get this updated.  It's used for logging messages so that if someone sends us a spammy long client string with a bunch of newlines and ascii stuff to advertise their ICO we don't end up with it clogging up our logging messages.
